go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -93,6 +94,97 @@ func pushDataset(ctx context.Context, destDB, srcDB datas.Database, tempTableDir
 	return err
 }
 
+type PushArg struct {
+	ds   datas.Dataset
+	db   datas.Database
+	hash hash.Hash
+}
+
+type AsyncPushOnWriteHook struct {
+	out io.Writer
+	ch  chan PushArg
+}
+
+const asyncPushBufferSize = 100
+
+var _ datas.CommitHook = (*AsyncPushOnWriteHook)(nil)
+
+// NewAsyncPushOnWriteHook creates a AsyncReplicateHook
+func NewAsyncPushOnWriteHook(ctx context.Context, destDB *DoltDB, tmpDir string) *AsyncPushOnWriteHook {
+	ch := make(chan PushArg, asyncPushBufferSize)
+
+	var newHeads = make(map[string]PushArg, asyncPushBufferSize)
+	go func() error {
+		defer close(ch)
+		for {
```

Comment:
This loop will only exit when ch is closed.  It won't exit if the context is cancelled, and this routine itself is responsible for closing the channel on exit.  So I don't think this routine will ever exit.

In most cases the writer should be responsible for closing the channel.